panther_hardware_interfaces/src/panther_imu.cpp

```diff
@@ -0,0 +1,513 @@
+  params_.cc_t0 = hardware_interface::stod(cc_t0);
+  params_.cc_t1 = hardware_interface::stod(cc_t1);
+  params_.cc_t2 = hardware_interface::stod(cc_t2);
+  params_.cc_t3 = hardware_interface::stod(cc_t3);
+  params_.cc_t4 = hardware_interface::stod(cc_t4);
+  params_.cc_t5 = hardware_interface::stod(cc_t5);
+}
+
+void PantherImuSensor::ReadMadgwickFilterParams()
+{
+  const auto gain = info_.hardware_parameters.at("gain");
+  const auto zeta = info_.hardware_parameters.at("zeta");
+  const auto mag_bias_x = info_.hardware_parameters.at("mag_bias_x");
+  const auto mag_bias_y = info_.hardware_parameters.at("mag_bias_y");
+  const auto mag_bias_z = info_.hardware_parameters.at("mag_bias_z");
+  const auto use_mag = info_.hardware_parameters.at("use_mag");
+  const auto stateless = info_.hardware_parameters.at("stateless");
+  const auto remove_gravity_vector = info_.hardware_parameters.at("remove_gravity_vector");
+
+  CheckMadgwickFilterWorldFrameParam();
+
+  params_.gain = hardware_interface::stod(gain);
+  params_.zeta = hardware_interface::stod(zeta);
+  params_.mag_bias_x = hardware_interface::stod(mag_bias_x);
+  params_.mag_bias_y = hardware_interface::stod(mag_bias_y);
+  params_.mag_bias_z = hardware_interface::stod(mag_bias_z);
+  params_.use_mag = hardware_interface::parse_bool(use_mag);
+  params_.stateless = hardware_interface::parse_bool(stateless);
+  params_.remove_gravity_vector = hardware_interface::parse_bool(remove_gravity_vector);
+}
+
+void PantherImuSensor::CheckMadgwickFilterWorldFrameParam()
+{
+  const auto world_frame = info_.hardware_parameters.at("world_frame");
+
+  auto it = kImuWorldFramesMap.find(world_frame);
+  if (it == kImuWorldFramesMap.end()) {
+    throw std::runtime_error(
+      "The parameter world_frame was set to invalid value. "
+      "Valid values are 'enu', 'ned' and 'nwu'. Setting to 'enu'.");
+  }
+  world_frame_ = it->second;
+}
+
+void PantherImuSensor::SetInitialValues()
+{
+  imu_sensor_state_.resize(
+    info_.sensors[0].state_interfaces.size(), std::numeric_limits<double>::quiet_NaN());
+}
+
+void PantherImuSensor::Calibrate()
+{
+  if (spatial_ == nullptr) {
+    throw std::runtime_error("Imu hardware is not active yet!");
+  }
+
+  spatial_->zero();
+  // The API call returns directly, so we "enforce" the recommended 2 sec
+  // here. See: https://github.com/ros-drivers/phidgets_drivers/issues/40
+
+  // FIXME: Ideally we'd use an rclcpp method that honors use_sim_time here,
+  // but that doesn't actually exist.  Once
+  // https://github.com/ros2/rclcpp/issues/465 is solved, we can revisit this.
+  std::this_thread::sleep_for(std::chrono::seconds(2));
+}
+
+bool PantherImuSensor::IsParamDefined(const std::string & param_name)
+{
+  return info_.hardware_parameters.find(param_name) != info_.hardware_parameters.end();
+}
+
+void PantherImuSensor::ConfigureCompassParams()
+{
+  if (
+    IsParamDefined("cc_mag_field") and IsParamDefined("cc_offset0") and
```

Comment:
Could we either overload `IsParamDefined` or introduce `AreParamsDefined` to check multiple parameters simultaneously, for instance, by passing an `std::unordered_set` or a similar container?

Reply:
Great idea!